### ISSUE TITLE
[docking] Window is movable even if the mouse is not directly on top of the window

### ISSUE THREAD

**Issue by garyng** [mellinoe/ImGui.NET#156]
_I fat-fingered and hit submit issue without filling in the issue details, sorry for that!_

In the [docking](https://github.com/mellinoe/ImGui.NET/tree/docking) branch, the window is moveable even if the mouse is not directly on top of the window:

![2020-03-24_21-23-44](https://user-images.githubusercontent.com/4453664/77430344-fca10100-6e15-11ea-93a2-a8916ef6997c.gif)

**Comment by garyng** [comment]
I think this is probably because of the imgui window is still receiving mouse input events even if the window is not in focus:

![2020-03-24_21-43-52](https://user-images.githubusercontent.com/4453664/77432252-c7e27900-6e18-11ea-92d7-381522bd014e.gif)

**Comment by AyphixLLC** [comment]
@garyng Windows 10 maintains some sort of focus even if the window isn't in focus, that's why you can scroll up and down on windows that aren't in focus. I'm not sure if or how this effects ImGui.

**Comment by garyng** [comment]
@AyphixLLC I think that only applies to scrolling right? 
The original imgui doesn't have this problem though.

**Comment by garyng** [comment]
Is this because of the use of `SDL_GetGlobalMouseState`? The docs says
> This can be useful if you need to track the mouse **outside** of a specific window

@mellinoe Any idea?

**Comment by garyng** [comment]
ImGui.NET: the window still get "hovered" event even if the viewport (?) is not focused

![2020-03-29_15-46-22](https://user-images.githubusercontent.com/4453664/77843958-a0066300-71d4-11ea-9bc7-06cbe3003154.gif)

Original imgui:
![2020-03-29_15-49-41](https://user-images.githubusercontent.com/4453664/77844004-f8d5fb80-71d4-11ea-9964-f8a34d3b2737.gif)

**Comment by sqrMin1** [comment]
It would be nice to know what's causing this and it would be nice to know a satisfying solution.

I noticed that the XNA sample project (which comes with this repo) still has this bug.
But the Veldrid sample (also from this repo) works like it's meant to be.

This is not directly a bug with the docking branch, but rather a general one.

I just tested it with the current master branch.

**Comment by sqrMin1** [comment]
Okay, I think that I found a solution to the problem and I created this gist here, which fixes the **ImGuiRenderer.cs**:
https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94

I simply subscribed to the [Activated & Deactivated](https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94#file-imguirenderer-cs-L52-L53) events from the MonoGame Game class and setting an [_inputPossible](https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94#file-imguirenderer-cs-L42) boolean field [correspondingly](https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94#file-imguirenderer-cs-L70-L78).

In the [UpdateInput](https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94#file-imguirenderer-cs-L231) prototype I check for the _inputPossible condition.

This happens right before [ImGui.NewFrame()](https://gist.github.com/sqrMin1/7449ffb818f790678c979c6085540b94#file-imguirenderer-cs-L137-L139) as the [official ImGui FAQ](https://github.com/ocornut/imgui/blob/master/docs/FAQ.md#q-how-can-i-tell-whether-to-dispatch-mousekeyboard-to-dear-imgui-or-to-my-application) recommends.

This FAQ also says: "Important: you should always pass your mouse/keyboard inputs to Dear ImGui[...]", which my solution kinda prevents. But this is eaxactly what I want in certain situations, despite that the FAQ states: "This is because e.g. we need to detect that you clicked in the void to unfocus its own windows, and other reasons."

My tests doesn't give bad results so far.

However, if you work with WindowsForms (without a game class) and ImGui.NET, you need to implement the above mentioned steps in every form (subscribe to Activated and Deactivated events).

This is BTW also how the official MonoGame.Framework handles this:

1.  [Activated & Deactivated event subscription](https://github.com/MonoGame/MonoGame/blob/5b040f6512c7acc79c7168f7df88ae6de4cb2529/MonoGame.Framework/Platform/Windows/WinFormsGameWindow.cs#L165-L166)
2. [Handle Input correspondingly](https://github.com/MonoGame/MonoGame/blob/5b040f6512c7acc79c7168f7df88ae6de4cb2529/MonoGame.Framework/Platform/Windows/WinFormsGameWindow.cs#L240-L258)

You just need to close the gap between the ImGuiRenderer and MonoGame or rather creating the bridge.

If there is a better solution I will be happy to hear it.

Cheers

**Comment by mellinoe** [comment]
@sqrMin1 The original issue here is about the multi-viewport Veldrid renderer. Are you seeing a similar problem in the MonoGame sample? I can't see any issue there, but I can repro the original one.

**Comment by sqrMin1** [comment]
I thought it could be a general problem, because I also got this bug in the MonoGame sample. Sorry for confusion.

I'm getting mouse-over effects on overlapping windows, but I couldn't directly click through with the current master.

![mousebug](https://user-images.githubusercontent.com/7989230/112546561-d9f35080-8db9-11eb-81d4-6102ab7ab2dc.gif)

However it was possible for me to reach "ImGui.IsItemHovered()" breakpoints:

![mousebug2](https://user-images.githubusercontent.com/7989230/112547857-a0234980-8dbb-11eb-83f4-c119c6a74f50.gif)

This is why I initially came to the conclusion that this could be connected with this issue. But it could be also a bug for it's own, so I opened #238.
